modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DotExpanderProcessor.java

```diff
@@ -0,0 +1,108 @@
+import org.elasticsearch.ingest.ConfigurationUtils;
+import org.elasticsearch.ingest.IngestDocument;
+import org.elasticsearch.ingest.Processor;
+
+import java.util.Map;
+
+public final class DotExpanderProcessor extends AbstractProcessor {
+
+    static final String TYPE = "dot_expander";
+
+    private final String path;
+    private final String field;
+
+    DotExpanderProcessor(String tag, String path, String field) {
+        super(tag);
+        this.path = path;
+        this.field = field;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void execute(IngestDocument ingestDocument) throws Exception {
+        String path;
+        Map<String, Object> map;
+        if (this.path != null) {
+            path = this.path + "." + field;
+            map = ingestDocument.getFieldValue(this.path, Map.class);
+        } else {
+            path = field;
+            map = ingestDocument.getSourceAndMetadata();
+        }
+
+        if (ingestDocument.hasField(path)) {
+            Object value = map.remove(field);
+            ingestDocument.appendFieldValue(path, value);
+        } else {
+            // check whether we actually can expand the field in question into an object field.
+            // part of the path may already exist and if part of it would be a value field (string, integer etc.)
+            // then we can't override it with an object field and we should fail with a good reason.
+            // IngestDocument#setFieldValue(...) would fail too, but the error isn't very understandable
+            for (int index = path.indexOf('.'); index != -1; index = path.indexOf('.', index + 1)) {
+                String partialPath = path.substring(0, index);
+                if (ingestDocument.hasField(partialPath)) {
+                    Object val = ingestDocument.getFieldValue(partialPath, Object.class);
+                    if ((val instanceof Map) == false) {
+                        throw new IllegalArgumentException("cannot expend [" + path + "], because [" + partialPath +
+                                "] is not an object field, but a value field");
+                    }
+                } else {
+                    break;
+                }
+            }
+            Object value = map.remove(field);
+            ingestDocument.setFieldValue(path, value);
+        }
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    String getPath() {
+        return path;
+    }
+
+    String getField() {
+        return field;
+    }
+
+    public static final class Factory implements Processor.Factory {
+
+        @Override
+        public Processor create(Map<String, Processor.Factory> processorFactories, String tag,
+                                Map<String, Object> config) throws Exception {
+            String field = ConfigurationUtils.readStringProperty(TYPE, tag, config, "field");
+            if (field.contains(".") == false) {
```

Comment:
since this is done once. should we be more restrictive here to actually have a proper path format? just a `.` is pretty flexible and would allow

```
hello.
.hello
```